enhancements/authentication/configuring-webhook-token-authenticators.md

```diff
@@ -0,0 +1,169 @@
+---
+title: configuring-webhook-token-authenticators
+authors:
+  - "@stlaz"
+reviewers:
+  - "@sttts"
+  - "@deads2k"
+approvers:
+  - "@deads2k"
+  - "@sttts"
+  - "@mfojtik"
+creation-date: 2019-12-10
+last-updated: 2019-12-10
+status: implementable
+see-also:
+  - "[oauth-apiserver enhancement](https://github.com/openshift/enhancements/pull/75)"
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# Configuring Webhook Token Authenticators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+## Summary
+
+The standard mechanism for external authentication integration is via `tokenreviews.authentication.k8s.io`
+from the kube-apiserver. We will support configuring the kube-apiserver this way and ensure it always
+works by using the same mechanism to drive the standard openshift oauth-server authentication integration.
+
+## Motivation
+
+Using the standard kube-apiserver extension mechanism will allow external integrations like Keycloak or
+any other Kubernetes authentication stack to integrate in a standard way.
+
+`webhookTokenAuthenticators` is a kube-apiserver configuration directive that can be used to
+configure an external authenticator which supports Kubernetes' token reviews (see
+[Webhook Token Authentication](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) for more details).
+By implementing this enhancement, we would significantly simplify any drop-in replacements
+for the integrated oauth-server, which might be helpful in cases where we would like to
+delegate authentication to a different OpenShift control plane, or direct authentication to an
+identity provider capable of handling the Kubernetes token reviews, such as Keycloak.
+
+### Goals
+
+1. Make it possible to configure an endpoint capable of performing kube token reviews
+2. Modify cluster-authentication-operator so that it configures authentication via the OpenShift OAuth
+   server by using the `webhookTokenAuthneticators` field instead of the current patching of the
+   kube-apiserver.
+
+### Non-Goals
+
+1. provide a partial authentication operator to any external authenticator.  Once the standard oauth-server
+   is disabled, authentication configuration is *entirely* the responsibility of the new authentication stack.
+2. allow both the built-in oauth-server and another authentication stack to be enabled at the same time.
+
+## Proposal
+
+### Implementation
+
+#### API
+
+`authentications.config.openshift.io/v1` already contains the `spec.webhookTokenAuthenticators`. It
```

Comment:
does that match 1:1 to the upstream flag + config?

Reply:
nvmd. It just referenes secrets which are eventually passed as configs to kube-apiserver.

Reply:
I added a note about it only being possible to add just one secret reference to match the upstream option

Reply:
if it didn't do anything before, just change it.